functional-samples/sample.tabcapture-recorder/offscreen.js

```diff
@@ -0,0 +1,79 @@
+// Copyright 2023 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+chrome.runtime.onMessage.addListener(async (message) => {
+  if (message.target === 'offscreen') {
+    switch (message.type) {
+      case 'start-recording':
+        startRecording(message.data);
+        break;
+      case 'stop-recording':
+        stopRecording();
+        break;
+      default:
+        throw new Error('Unrecognized message:', message.type);
+    }
+  }
+});
+
+let recorder;
+let data = [];
+
+async function startRecording(streamId) {
+  const media = await navigator.mediaDevices.getUserMedia({
+    audio: {
+      mandatory: {
+        chromeMediaSource: 'tab',
+        chromeMediaSourceId: streamId
+      }
+    },
+    video: {
+      mandatory: {
+        chromeMediaSource: 'tab',
+        chromeMediaSourceId: streamId
+      }
+    }
+  });
+
+  // Continue to play the captured audio to the user.
+  const output = new AudioContext();
+  const source = output.createMediaStreamSource(media);
+  source.connect(output.destination);
+
+  // Start recording.
+  recorder = new MediaRecorder(media, { mimeType: 'video/webm' });
+  recorder.ondataavailable = (event) => data.push(event.data);
+  recorder.onstop = () => {
+    const blob = new Blob(data, { type: 'video/webm' });
+    window.open(URL.createObjectURL(blob), '_blank');
+
+    // Clear state ready for next recording
+    recorder = undefined;
+    data = [];
+  };
+  recorder.start();
+
+  // Record current state in URL
+  window.location.hash = 'recording';
+}
+
+async function stopRecording() {
```

Comment:
Would it make sense to close the offscreen document here?

Reply:
I included some thoughts on this in the PR description. It would be nice, but it would require a lot of extra bulk since I would need to serialize the data and send it out of the offscreen document (object URLs are cleared when the associated page unloads). I think that would make the sample a lot harder to understand.

Reply:
Sorry, somehow I always forget to read the PR description on Github. 

Would it be possible to simply send `data` via message and then open the tab from the sw?

Reply:
We could, but Blob's aren't serializable so we would need to extract the data. I'm also not super keen on sending large recordings over the messaging APIs since I suspect that will eventually lead to issues. The alternative is storing the data somewhere but that would be a lot of logic (we would probably need something like IndexedDB).

Reply:
```
const blob = new Blob(data, { type: 'video/webm' });
```

I didn't mean sending the Blob, but `data` is a byte array which should be no problem to transmit, no? I also couldn't find any limitations regarding message size (this is mostly for my own edification as I'm curious about the limitations here). 

The reason I'm also asking: this looks like a problem that everyone has who's using our proposed workaround for background recording. But maybe I'm looking at this wrong and it most "real" cases the UX is completely different so that there's a clear point when to close the offscreen doc.